pkg/sync/rpc/server.go

```diff
@@ -1399,109 +1402,45 @@ func (s *SyncAgentServer) rmDisk(disk string) error {
 	return nil
 }
 
-// The APIs BackupAdd, BackupGet, Refresh, BackupDelete implement the CRUD interface for the backup object
-// The slice Backup.backupList is implemented similar to a FIFO queue.
-
-// BackupAdd creates a new backupList object and appends to the end of the list maintained by backup object
-func (b *BackupList) BackupAdd(backupID string, backup *replica.BackupStatus) error {
-	if backupID == "" {
-		return fmt.Errorf("empty backupID")
+func (s *SyncAgentServer) SnapshotHash(ctx context.Context, req *ptypes.SnapshotHashRequest) (*empty.Empty, error) {
+	if len(req.SnapshotNames) == 0 {
+		return nil, fmt.Errorf("snapshot name(s) is required")
 	}
 
-	b.Lock()
-	b.backups = append(b.backups, &BackupInfo{
-		backupID:     backupID,
-		backupStatus: backup,
-	})
-	b.Unlock()
-
-	if err := b.Refresh(); err != nil {
-		return err
+	size := s.snapshotHashingWorkerPool.GetSize()
+	if req.ConcurrentLimit > 0 && int(req.ConcurrentLimit) != size {
+		s.snapshotHashingWorkerPool.Resize(int(req.ConcurrentLimit))
 	}
 
-	return nil
-}
+	task := replica.NewSnapshotHashTask(req.SnapshotNames, req.Rehash)
 
-// BackupGet takes backupID input and will return the backup object corresponding to that backupID or error if not found
-func (b *BackupList) BackupGet(backupID string) (*replica.BackupStatus, error) {
-	if backupID == "" {
-		return nil, fmt.Errorf("empty backupID")
+	if err := s.SnapshotHashList.Delete(req.SnapshotNames[0]); err != nil {
+		return nil, errors.Wrapf(err, "failed to delete snapshot %v for from list", req.SnapshotNames[0])
 	}
 
-	if err := b.Refresh(); err != nil {
-		return nil, err
+	if err := s.SnapshotHashList.Add(req.SnapshotNames[0], task); err != nil {
+		return nil, errors.Wrapf(err, "failed to add snapshot %v for hashing", req.SnapshotNames[0])
 	}
 
-	b.RLock()
-	defer b.RUnlock()
+	s.snapshotHashingWorkerPool.Exec(task)
 
-	for _, info := range b.backups {
-		if info.backupID == backupID {
-			return info.backupStatus, nil
-		}
-	}
-	return nil, fmt.Errorf("backup not found %v", backupID)
+	return &empty.Empty{}, nil
 }
 
-// remove deletes the object present at slice[index] and returns the remaining elements of slice yet maintaining
-// the original order of elements in the slice
-func (*BackupList) remove(b []*BackupInfo, index int) ([]*BackupInfo, error) {
-	if b == nil {
-		return nil, fmt.Errorf("empty list")
-	}
-	if index >= len(b) || index < 0 {
-		return nil, fmt.Errorf("BUG: attempting to delete an out of range index entry from backupList")
+func (s *SyncAgentServer) SnapshotHashStatus(ctx context.Context, req *ptypes.SnapshotHashStatusRequest) (*ptypes.SnapshotHashStatusResponse, error) {
+	if req.SnapshotName == "" {
```

Comment:
Do we need to check this when it's also checked in `s.SnapshotHashList.Get`?

Reply:
Do we need to get the status one by one? Unlike the backup creation, the snapshot hash can be handled in parallel.

Reply:
You are right. Concurrent way is better. Thank you.

Reply:
Using waitgrounp instead, thank you,